src/qibolab/instruments/qm.py

```diff
@@ -581,7 +581,10 @@ def fetch_results(result, ro_pulses):
                 shots = handles.get(f"{serial}_shots").fetch_all().astype(int)
             else:
                 shots = None
-            results[pulse.qubit] = results[serial] = ExecutionResults.from_components(ires, qres, shots)
+            if average:
+                results[pulse.qubit] = results[serial] = AveragedResults.from_components(ires, qres)
+            else:
+                results[pulse.qubit] = results[serial] = ExecutionResults.from_components(ires, qres, shots)
```

Comment:
In this case I'm advocating for something different, but in another situation you could reduce duplication placing the `if` just on the RHS:
```suggestion
                results[pulse.qubit] = results[serial] = AveragedResults.from_components(ires, qres) if average else ExecutionResults.from_components(ires, qres, shots)
```

If `shots` were not there you could have also split the statement with `Results = AveragedResults if average else ExecutionResults`, and then a single `Results.from_components`.
Of course, it is possible even with `shots`, but you'd need another `if` for the argument list, and it'd start being more cumbersome again.

Reply:
Thanks you are right.